Fully serialize ROMClasses sent to JITServer

AlexeyKhrabrov opened the pull request:
Packed ROMClasses that JIT client sends to the server do not always include UTF8 strings due to interning. In the current implementation, JITServer requests missing strings from the client and caches them in the client session. To avoid these requests, in this PR the ROMClass is fully serialized with all the required strings that it refers to.

All the strings are written at the end of the serialized ROMClass in a fixed order that is determined by the ROMClass walk implementation, regardless of whether they are stored within the bounds of the original ROMClass or interned.

The following data is not used by JITServer and is excluded from the resulting packed ROMClass, with their SRPs set to 0:
- Intermediate class data.
- Strings that are part of inline method debug info. (The rest of method debug info data is stored in ROMMethod, and removing it would require shifting everything after it and updating SRPs, which is not easy to implement.)
Other strings that are guaranteed not to be used by JITServer can also be excluded in the future.

Note: this PR depends on #11330

Closes: #11329

AlexeyKhrabrov commented:
Here are some relevant message statistics from AcmeAir start. Before the change:
```
Type# #called           Max             Min             Mean            StdDev          Sum             TypeName
...
#0005    4465   83504.000000    1752.000000     9398.477044     12309.351652    41964200.000000         compilationRequest
...
#0034    9453   228.000000      28.000000       50.261293       18.909630       475120.000000           ResolvedMethod_getRemoteROMString
...
#0036    1959   54472.000000    504.000000      3457.278203     4850.623491     6772808.000000          ResolvedMethod_getRemoteROMClassAndMethods
...
#0236      10   88.000000       40.000000       66.800000       16.870750       668.000000              ClassInfo_getRemoteROMString
...
Total number of messages: 163088. Average number of messages per compilation:36.550426
```
After the change:
```
Type# #called           Max             Min             Mean            StdDev          Sum             TypeName
...
#0005    4481   81648.000000    1760.000000     9200.235662     12065.460703    41226256.000000         compilationRequest
...
#0035    1978   60168.000000    504.000000      4202.883721     5591.390639     8313304.000000          ResolvedMethod_getRemoteROMClassAndMethods
...
Total number of messages: 155864. Average number of messages per compilation:34.822163
```

There is a trade-off between reducing the number of messages and the increased size of ROMClasses stored at the server, which we can estimate based on message sizes: `getRemoteROMClassAndMethods + getRemoteROMString` before vs `getRemoteROMClassAndMethods` after. This gives an upper boundary of ~15% increase (it doesn't account for the memory overhead of the hash table that stores cached remote strings). However, I didn't observe any measurable changes in overall JITServer memory consumption in this experiment. The size of packed ROMClasses can be reduced if we identify more strings that are never accessed by the JIT and can be safely excluded during serialization. Also, this won't really matter if we implement storing a single copy of ROMClasses at JITServer shared by multiple clients.

AlexeyKhrabrov commented:
Using the ROMClass walk adds a bit of overhead (< 2% of client JIT CPU time) because it visits every field in every ROM structure. This overhead can be eliminated by either adding an option to only visit strings in `allSlotsInROMClassDo()` (which means more changes to VM code), or by implementing a separate string-only walk (which means more maintenance when adding/changing ROMClass data).

AlexeyKhrabrov commented:
Marking as draft until the prerequisite #11330 is merged, but otherwise it should be ready. @mpirvu Could you please review?

AlexeyKhrabrov commented:
Prerequisite PR has been merged, ready for review now.

AlexeyKhrabrov commented:
Rebased and incremented the JITServer protocol version (it was updated to the same value by a different PR).

mpirvu commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+                      "Missed strings in ROMClass %.*s UTF8 section: %zu != %zu",
+                      name->length, name->data, ctx.utf8SectionSize, end - ctx.utf8SectionStart);
+      end = (const uint8_t *)OMR::alignNoCheck((uintptr_t)end, sizeof(uint64_t));
+      TR_ASSERT_FATAL(end == classEnd, "UTF8 section not stored at the end of ROMClass %.*s: %p != %p",
```
Why does `end` need to be the same as `classEnd`?
`end` is what we computed with interned and non-interned strings.
`classEnd` is `romClass+romClass->romSize` which I thought it did not include the interned strings

AlexeyKhrabrov replied:
`end` is computed with non-interned strings only.

mpirvu replied:
Right

mpirvu commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+   }
+
+// Array ROMClasses have a different layout (see runtime/vm/romclasses.c):
+// they all share the same SRP array of interfaces, which breaks the generic
+// packing implementation. Instead, we manually pack all the data stored
+// outside of the ROMClass header: class name, superclass name, interfaces.
+// This function returns the total size of the packed array ROMClass.
```
Does this mean that for array ROMClasses we may send interface names multiple times?

AlexeyKhrabrov replied:
For a given array ROMClass (or any ROMClass) we send all the strings it refers to (a single copy of each string). Overall, if we consider all the ROMClasses for a given client, there will be multiple copies of some strings (ones that were interned at the client). But it's the same for the current implementation - remote strings are retrieved and cached for each class separately. Deduplicating the strings on the server side would be non-trivial.

mpirvu reviewed (commented):
Looks good overall. I have some questions for my own understanding.
The danger here is that the VM may change some of the assumptions about ROMClass structure that his implementation makes.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+// Copies a UTF8 string with padding and returns its total size
 static size_t
-getPackedUTF8Size(const J9UTF8 *str, const J9ROMClass *romClass)
+copyUTF8(J9UTF8 *dst, const J9UTF8 *src)
```
Correct. Will add a comment.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+   static_assert(sizeof(*src) == 2, "UTF8 header is not 2 bytes large");
+   if (!OMR::alignedNoCheck(size, sizeof(*src)))
+      dst->data[src->length] = '\0';
+   return OMR::alignNoCheck(size, sizeof(*src));
```
It returns the padded size, i.e. getUTF8Size(src) == getUTF8Size(dst), but it's different from J9UTF8_TOTAL_SIZE(dst). I'll change it to return getUTF8Size(dst).

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+      return;
+      }
+
+   ctx.strToOffsetMap.insert(it, { str, skip ? (size_t)-1 : ctx.stringsSize });
```
Yes, it's a bit faster since it avoids a hash lookup.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+//
+// Packing involves 2 passes over all the strings that a ROMClass refers to:
+// 1. Compute total size and map original strings to their locations in the packed ROMClass.
+// 2. Copy each original string to its location in the packed ROMClass.
```
In general, no. I will add a comment about that.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
-      romMethod = nextROMMethod(romMethod);
+      TR_ASSERT_FATAL(icData + romClass->intermediateClassDataLength == (uint8_t *)romClass + romClass->romSize,
+                      "Intermediate class data not stored at the end of ROMClass %.*s", name->length, name->data);
+      totalSize -= romClass->intermediateClassDataLength;
```
True, but this way the code is a bit simpler.

AlexeyKhrabrov commented on runtime/compiler/control/JITServerHelpers.cpp:
```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+   size_t totalSize = origRomSize;
+
+   // Remove intermediate class data (not used by JIT)
+   uint8_t *icData = J9ROMCLASS_INTERMEDIATECLASSDATA(romClass);
```
Not sure, I haven't explored enough VM code to fully understand it yet. But it looks like the JIT never accesses it, so it should be safe to exclude it if it's present.

AlexeyKhrabrov commented:
I've addressed the review comments. Let me know if anything else needs to be clarified. The danger of future VM changes breaking our assumptions about ROMClass layout is somewhat mitigated by checking them at runtime with fatal assertions.

I've also tried a slightly different implementation that remembers SRP slots visited during the 1st pass instead of performing a full ROMClass walk during the 2nd pass, but it didn't show a measurable reduction in client CPU overhead. The only way to optimize this implementation would be to have a faster version of the ROMClass walk that only visits string SRP slots.

mpirvu commented:
jenkins test sanity plinuxjit,xlinujit,zlinuxjit jdk11

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
Tests revealed some errors:
```
===============================================
Running test JLM_Tests_interface_0 ...
===============================================
JLM_Tests_interface_0 Start Time: Wed Dec  9 20:13:46 2020 Epoch Time (ms): 1607562826138
variation: NoOptions
JVM_OPTIONS: -XX:+UseJITServer 
JITServer is currently a technology preview. Its use is not yet supported
JVMSHRC023E Cache does not exist
Assertion failed at /home/jenkins/workspace/Build_JDK11_ppc64le_linux_jit_Personal/openj9/runtime/compiler/control/JITServerHelpers.cpp:317: ctx._utf8SectionSize == end - ctx._utf8SectionStart
VMState: 0x0005ffff
	Missed strings in ROMClass java/lang/System UTF8 section: 4476 != 4508
compiling java/lang/System.getEncoding(I)Ljava/lang/String; at level: cold
```

AlexeyKhrabrov commented:
I found at least one issue that might explain the test failures. ROMClass walker doesn't recognize constant dynamic CP entries - it stops walking the constant pool if it encounters one, so for any class with a constant dynamic CP entry we risk skipping a part of its constant pool (not sure if `java.lang.System` is one of them though). I'll submit a PR with the walker fix, and mark this one as draft until it's merged and until I make sure that this is actually the cause of the test failures.

AlexeyKhrabrov commented:
@mpirvu The required VM changes have been merged, this should be good to go now. I've also rebased and updated copyright years and JITServer protocol version.

mpirvu reviewed (approved):
LGTM

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

mpirvu commented:
I will do a performance sniff before merging this one.

mpirvu commented:
Performance looks stable. The CPU consumption at the client is practically unchanged which means that the reduction in number of messages has ben offset by an overhead increase due to packing ROMClasses. Average number of messages has been decreased somewhat from 53.5 to 49.4.
Footprint seems to have increased ~1%, but maybe that's just fluctuation.
AcmeAir performance before the change:
```
Throughput      avg=13779.89    min=13327.60    max=13982.60    stdDev=226.3    maxVar=4.91%    confInt=0.95%   samples=10
Intermediate results:
Run 0   794.2   13214.9 13344.1 13454.1 Avg=13454       CPU=9883 ms  Footprint=282124 KB
Run 1   761.1   13741.0 13559.2 13914.9 Avg=13915       CPU=5178 ms  Footprint=267808 KB
Run 2   756.2   13560.0 13678.6 13327.6 Avg=13328       CPU=4909 ms  Footprint=265732 KB
Run 3   772.3   13736.3 13803.3 13677.1 Avg=13677       CPU=4663 ms  Footprint=265212 KB
Run 4   765.4   13879.6 13912.6 13778.5 Avg=13778       CPU=4602 ms  Footprint=264564 KB
Run 5   774.0   13754.8 13833.6 13882.5 Avg=13882       CPU=4621 ms  Footprint=264216 KB
Run 6   766.3   13700.3 13845.9 13877.0 Avg=13877       CPU=4526 ms  Footprint=267076 KB
Run 7   768.9   13728.0 13799.3 13931.2 Avg=13931       CPU=4495 ms  Footprint=265104 KB
Run 8   777.2   13851.8 13947.4 13973.4 Avg=13973       CPU=4372 ms  Footprint=264612 KB
Run 9   770.8   13761.7 13867.5 13982.6 Avg=13983       CPU=4343 ms  Footprint=265192 KB
CompTime        avg=5159.20     min=4343.00     max=9883.00     stdDev=1678.3   maxVar=127.56%  confInt=18.86%  samples=10
Footprint       avg=267164.00   min=264216.00   max=282124.00   stdDev=5376.7   maxVar=6.78%    confInt=1.17%   samples=10
```
After the change:
```
Throughput      avg=13736.92    min=13470.70    max=13948.80    stdDev=132.6    maxVar=3.55%    confInt=0.56%   samples=10
Intermediate results:
Run 0   791.8   13353.1 13523.0 13470.7 Avg=13471       CPU=9930 ms  Footprint=282352 KB
Run 1   769.2   13752.2 13804.8 13750.0 Avg=13750       CPU=4875 ms  Footprint=268572 KB
Run 2   772.9   13803.0 13917.6 13948.8 Avg=13949       CPU=4863 ms  Footprint=267376 KB
Run 3   772.2   13744.0 13794.0 13802.2 Avg=13802       CPU=4812 ms  Footprint=269032 KB
Run 4   776.5   13654.3 13718.5 13725.8 Avg=13726       CPU=4778 ms  Footprint=266980 KB
Run 5   767.2   13727.4 13791.5 13871.2 Avg=13871       CPU=4625 ms  Footprint=268164 KB
Run 6   770.4   13480.1 13631.3 13645.2 Avg=13645       CPU=4746 ms  Footprint=269140 KB
Run 7   773.7   13637.6 13750.6 13643.7 Avg=13644       CPU=4526 ms  Footprint=267912 KB
Run 8   770.3   13581.7 13750.7 13788.2 Avg=13788       CPU=4498 ms  Footprint=266996 KB
Run 9   771.9   13590.7 13729.7 13723.4 Avg=13723       CPU=4395 ms  Footprint=267408 KB
CompTime        avg=5204.80     min=4395.00     max=9930.00     stdDev=1668.4   maxVar=125.94%  confInt=18.58%  samples=10
Footprint       avg=269393.20   min=266980.00   max=282352.00   stdDev=4619.8   maxVar=5.76%    confInt=0.99%   samples=10
```
One thing that caught my eye was an increase in the number `compilationAotClassReloFailure` from 72 to 305 (based on just 2 samples). If this holds, we may want to look into it.

mpirvu commented:
I had more runs and the number of compilation failures is highly variable.
This PR is ready to be merged.
